python/pyarrow/tests/test_udf.py

```diff
@@ -504,3 +504,132 @@ def test_input_lifetime(unary_func_fixture):
+            else:
+                return "count: " + str(self.non_null)
+
+    def init():
+        state = State(0)
+        return state
+
+    def consume(ctx, x):
+        if isinstance(x, pa.Array):
+            non_null = pc.sum(pc.invert(pc.is_nan(x))).as_py()
+        elif isinstance(x, pa.Scalar):
+            if x.as_py():
+                non_null = 1
+        non_null = non_null + ctx.state.non_null
+        return State(non_null)
+
+    def merge(ctx, other_state):
+        merged_state_val = ctx.state.non_null + other_state.non_null
+        return State(merged_state_val)
+
+    def finalize(ctx):
+        return pa.array([ctx.state.non_null])
+
+    func_name = "simple_count"
+    unary_doc = {"summary": "count function",
+                 "description": "test agg count function"}
+
+    pc.register_scalar_aggregate_function(init,
+                                          consume,
+                                          merge,
+                                          finalize,
+                                          func_name,
+                                          unary_doc,
+                                          {"array": pa.int64()},
+                                          pa.int64())
+
+    assert pc.call_function(func_name, [pa.array(
+        [10, 20, None, 30, None, 40])]) == pa.array([4])
+
+
+def test_aggregate_udf_with_custom_state_multi_attr():
+    class State:
+        def __init__(self, non_null=0, null=0):
+            self._non_null = non_null
+            self._null = null
+
+        @property
+        def non_null(self):
+            return self._non_null
+
+        @non_null.setter
+        def non_null(self, value):
+            self._non_null = value
+
+        @property
+        def null(self):
+            return self._null
+
+        @null.setter
+        def null(self, value):
+            self._null = value
+
+        def __repr__(self):
+            if self._non_null is None:
+                return "no values stored"
+            else:
+                return "non_null: " + str(self.non_null) \
+                    + ", null: " + str(self.null)
+
+    def init():
+        state = State(0, 0)
+        return state
+
+    def consume(ctx, x):
+        null = 0
+        non_null = 0
+        if isinstance(x, pa.Array):
+            non_null = pc.sum(pc.invert(pc.is_nan(x))).as_py()
+            null = len(x) - non_null
+        elif isinstance(x, pa.Scalar):
+            if x.as_py():
+                non_null = 1
+            else:
+                null = 1
+        non_null = non_null + ctx.state.non_null
+        return State(non_null, null)
+
+    def merge(ctx, other_state):
+        merged_st_non_null = ctx.state.non_null + other_state.non_null
+        merged_st_null = ctx.state.null + other_state.null
+        return State(merged_st_non_null, merged_st_null)
+
+    def finalize(ctx):
+        print(ctx.state)
+        return pa.array([ctx.state.non_null, ctx.state.null])
```

Comment:
An aggregate UDF must always return an array of length 1.  Do you perhaps want something like...

```
return pa.array([{"non_null": ctx.state.non_null, "null": ctx.state.null}])
```